contracts/SwapAndBridge.sol

```diff
@@ -103,8 +104,13 @@ abstract contract SwapAndBridgeBase is Lockable, MultiCaller {
         // user on this contract, a malicious actor could call transferFrom to steal the user's tokens.
         if (!allowedSelectors[bytes4(routerCalldata)]) revert InvalidFunctionSelector();
 
-        // Pull tokens from caller into this contract.
-        _swapToken.safeTransferFrom(msg.sender, address(this), swapTokenAmount);
+        // Pull tokens from caller into this contract. This contract may already have tokens if, for example,
+        // a user sent the contract tokens and subsequently called `swapAndBridge`, so only pull funds if
+        // the contract needs to.
```

Comment:
This is a nice feature but curious why you're adding it to this PR, should it be in a separate PR?

Reply:
I added it here since `receiveWithAuthorization` should send tokens to the contract. So ideally, whenever `swapAndBridgeWithAuthorization` is called, `transferFrom` should never be called, and for the normal and permit swap and bridge, `transferFrom` should be called.

Reply:
I see. Part of me thinks we should separate the token transfer logic from the swapping logic. Perhaps we should add a separate internal function called `_pullTokens` or something which has different behavior for `transferFrom` versus `permit` versus `receiveWithAuthorization `. What do you think?

Reply:
Yes, I think this will simplify things. I added a `_pullTokens` function in this commit: https://github.com/across-protocol/contracts/pull/751/commits/5f7ee7ba84f4b759ba9525a7d49dfd3455b434dc. I think the good thing about doing this is now all we really care about is whether the contract has sufficient balance to make the deposit/swap, and the way it gets the tokens is irrelevant.

Reply:
exactly

Reply:
so actually what I was suggesting is something more along the lines of removing this pulling logic from this function and leaving it the caller of `_swapAndBridge` to pull tokens before entering the function. `swapAndBridgeWithPermit`, `swapAndBridgeWithAuthorization` and `swapAndBridge` each have their own ways of pulling the user's funds. They can perform any checks they want to after pulling, and then they can call `_swapAndBridge`. What do you think of that? Am open to other ideas

Reply:
Ah I see. So basically, leave this part to function in case we were to add more ways of depositing/authenticating. Makes sense. I'm just jumping around these different options. Under the impression that this contract will be modified more times in the future, I think it probably would be better this way. https://github.com/across-protocol/contracts/pull/751/commits/7d67ab56d693289fd6d71933b4a5084503139a07